classic_cp/kidneyexchange/kidneyexchange.jl

```diff
@@ -8,98 +8,124 @@ end
+    v[i] = i => pair i does not participate in any cycle
+"""
+function solve_kidneyexchange_vector(filename::String)
 
-    c = filter(e -> !isempty(e), c)
+    inputData = getInputData(filename)
+    c = inputData.compatibilities
+    trailer = SeaPearl.Trailer()
+    model = SeaPearl.CPModel(trailer)
 
-    for i in 1:length(c)
-        c[i] = map(e -> pairsEquivalence[e], c[i])
-        push!(c[i], i)
+    ### Attempting to reduce the instance ###    
+    canReduce, pairsEquivalence, reduced_c = reduce_instance(inputData)
+    if canReduce
+        model.adhocInfo = pairsEquivalence
+        c = reduced_c
     end
-
-    println("Instance reduction: original size "*string(n)*", new size "*string(current_n))
-    println()
-    n = current_n
+    n = length(c)
     
     ### Variable declaration ###    
+
+    # x[i] = j => pair j receive a kidney from pair i
+    # x[i] = i => pair i ∉ cycles
     x = Vector{SeaPearl.AbstractIntVar}(undef, n) 
-    x_offset = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+
+    # "fake" variable to know if a pair participate in a cycle
+    index = Vector{SeaPearl.AbstractVar}(undef, n) 
+
+    # "view" variable to be able to use the "ReifiedInSet" constraint
+    # xEqualIndex_Bool[i] = true => i ∉ cycle
+    # xEqualIndex_Bool[i] = false => i ∈ cycle
+    xEqualIndex_Bool = Vector{SeaPearl.AbstractBoolVar}(undef, n)  
+
+    # "view" variable to be able to use the "ReifiedInSet" constraint
+    # xNotEqualIndex_Int[i] = 1 => i ∉ cycle
+    # xNotEqualIndex_Int[i] = 0 => i ∈ cycle
+    xNotEqualIndex_Int = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+
+    # "fake" variable to know if xNotEqualIndex_Int == 0
+    zeroSet = SeaPearl.IntSetVar(0, 0, "zeroSet", model.trailer)
+    # require the unique value to ensure that zeroSet is bounded
+    SeaPearl.require!(zeroSet.domain, 0)
 
     for i = 1:n
         x[i] = SeaPearl.IntVar(1, n, "x_"*string(i), model.trailer)
         SeaPearl.addVariable!(model, x[i]; branchable=true)
-        x_offset[i] = SeaPearl.IntVarViewOffset(x[i], -i, x[i].id*"-"*string(i))
+
+        index[i] = SeaPearl.IntSetVar(i, i, "index_"*string(i), model.trailer)
+        # require the unique value to ensure that isbound(index[i]) return true
+        SeaPearl.require!(index[i].domain, i)
+        
+        xEqualIndex_Bool[i] = SeaPearl.BoolVar("xEqualIndex_Bool_"*string(i), model.trailer)
+        xNotEqualIndex_Int[i] = SeaPearl.IntVar(0, 1,"xNotEqualIndex_Int_"*string(i), model.trailer)
     end
 
     ### Constraints ###
     for i = 1:n
+        #Fix xEqualIndex_Bool
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(x[i], index[i], xEqualIndex_Bool[i], trailer))
+
+        #Fix xNotEqualIndex_Int
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(xNotEqualIndex_Int[i], zeroSet, xEqualIndex_Bool[i], trailer))
+
+        #Add incompatibilities to the model
         for j in 1:n
-            if j ∉ c[i]
+            if j ∉ c[i] && i !=j
```

Comment:
So we do not consider the case where a pair can be paired to itself?

Reply:
Sorry, I understood later the underlying process using the `xEqualIndex_Bool` variables.